### PR TITLE
 BZ 1624014: Edits after NFS note feedback

### PR DESCRIPTION
**Pull request by bmcelvee**
Follow-up to https://github.com/openshift/openshift-docs/pull/11906. 

@sferich888 I'm continuing edits from Brice's original PR here.

### CODE REVIEW

File: scaling_performance/optimizing_storage.adoc [hunk 1]
```diff
@@ -135,11 +137,6 @@ In a scaled/HA {product-title} registry cluster deployment:
 * All NAS storage (excluding {gluster-native}/{gluster-external} GlusterFS as it uses an object storage interface) are
 not recommended for {product-title} Registry cluster deployment with production workloads.
 
-[WARNING]
-====
-include::install_config/topics/NFS_note.adoc[]
-====
-
 [[metrics]]
 ==== Metrics
```
Comment (bmcelvee) [inline scaling_performance/optimizing_storage.adoc]: sferich888 5 days ago Member

Overall this document needs to be looked at more closely.
Under Logging and Metrics (subsections) we have "warning" notes: "Corruption may occur when using NFS to back hosted logging with production workloads."

    We should remove these (and align the warning with other messaging (in other places) and to the table above.
        The table above says we support RWO, and RWX (as configurable options for both logging and metrics).
        In locations (other docs) we call out NFS backed file systems should not be used for various reasons (we should link to these notes and callouts).

Under the Logging subsection, we have the note:

    "NFS does not guarantee read-after-write consistency and is not recommended for applications which require it."
        This should be removed, as not every application requires this (so it may not be relevant).
    "Applications that depend on writing to the same, shared NFS export may experience issues with production workloads."
        We should update this to state:
            Applications are responsible for knowing and understanding the storage requirements for their application. It is on the application/development/deployment team to understand how an application work with the backed/provided storage, to ensure that issues do not occur with an application scales or interacts with the storage layer.

(Copied from original PR for tracking)

**Comment by bmcelvee** [comment]
@vikram-redhat I updated/closed the original PR, and am continuing the edits here.

File: scaling_performance/optimizing_storage.adoc (hunk 1 above)
Comment (bmcelvee) [reply scaling_performance/optimizing_storage.adoc]: Sorry for the wonky formatting in the copy of the previous comment. There was a lot to the comment, so I'm attempting to address it all here: 

- "Under Logging and Metrics (subsections) we have 'warning' notes: 'Corruption may occur when using NFS to back hosted logging with production workloads.' We should remove these (and align the warning with other messaging (in other places) and to the table above. "
-- Done

- "The table above says we support RWO, and RWX (as configurable options for both logging and metrics). In locations (other docs) we call out NFS backed file systems should not be used for various reasons (we should link to these notes and callouts)."
-- Please provide further guidance. Are the links necessary here?

- "Under the Logging subsection, we have the note: 'NFS does not guarantee read-after-write consistency and is not recommended for applications which require it."
-- Removed
- "'Applications that depend on writing to the same, shared NFS export may experience issues with production workloads.' We should update this to state: Applications are responsible for knowing and understanding the storage requirements for their application. It is on the application/development/deployment team to understand how an application work with the backed/provided storage, to ensure that issues do not occur with an application scales or interacts with the storage layer."
-- Done. (with minor edits)

**Comment by bmcelvee** [comment]
@sferich888 PTAL

**Comment by sferich888** [comment]
In the table where we denoted what types of volumes you can use with each components, we outline what's possible, we should connect this to what's recommend. 

For example with metrics, it says you can use RWO or RWX (single shared volume), however in RWX situations, this is an anti-pattern - we should call this out. With logging, and RWX situation, it's likely that the underlying file locking mechanisms (again ymmv depending on storage provider) this it's better to use 1 volume per DC.

**Comment by sjwalter2** [comment]
Actually for logging its a *requirement* to have one volume per logging-es DC. This is because each DC represents a node of an Elasticsearch cluster and has its own state information, indices, etc. By design they should be unique volumes. Just to clarify!

**Comment by sferich888** [comment]
It too is then an anti-pattern to use RWX storage for logging. We might want to call that out in the table for those components.

File: scaling_performance/optimizing_storage.adoc [hunk 2]
```diff
@@ -106,9 +106,30 @@ The following table summarizes the recommended and configurable storage technolo
...
 ====
 
+[IMPORTANT]
+====
+For metrics you can use RWO or RWX as a single shared volume. However, using RWX
+is an anti-pattern.
+
+For logging, you are required to have one volume per `logging-es` DC. However,
+using RWX is an anti-pattern. 
+====
```
Comment (bmcelvee) [inline scaling_performance/optimizing_storage.adoc]: Thank you, @sferich888 and @sjwalter2! I added a warning following the table. It needs a bit more word-smithing, but is this along the lines of what you're looking for?

For the  "However, using RWX is an anti-pattern." clause, would it be better written, "However, using RWX can cause an anti-pattern."?

Comment (vikram-redhat) [reply scaling_performance/optimizing_storage.adoc]: @sferich888 @sjwalter2 - *bump*. We want to close this as soon as possible.

File: scaling_performance/scaling_cluster_metrics.adoc [hunk 3]
```diff
@@ -86,10 +86,10 @@ storage size requirements for week, as in the values above.
 Because {product-title} metrics uses the Cassandra database as a datastore for
 metrics data, if `USE_PERSISTENT_STORAGE=true` is set during the metrics set up
 process, `PV` will be on top in the network storage, with NFS as the default.
-However, using network storage in combination with Cassandra is not recommended,
-as per the
-link:https://docs.datastax.com/en/landing_page/doc/landing_page/planning/planningAntiPatterns.html[Cassandra
-documentation].
+However, using network storage in combination with Cassandra is not recommended.
+
+If you use a Cassandra database as a datastore for metrics data, see Cassandra
```
Comment (lpsantil) [inline scaling_performance/scaling_cluster_metrics.adoc]: I don't recommend we remove a worthy source material link.

File: scaling_performance/optimizing_storage.adoc [hunk 4]
```diff
@@ -173,8 +198,10 @@ Application use cases vary from application to application, as described in the
 
 * Storage technologies that support dynamic PV provisioning have low mount time latencies, and are not tied
 to nodes to support a healthy cluster.
-* NFS does not guarantee read-after-write consistency and is not recommended for applications which require it.
-* Applications that depend on writing to the same, shared NFS export may experience issues with production workloads.
+* Application developers are responsible for knowing and understanding the storage
```
Comment (lpsantil) [inline scaling_performance/optimizing_storage.adoc]: Often, Application Developers have little or no say in their storage backends.  They take what they are given, for better or for worse.  It also intends to force Application Developers to learn about storage.  A worthy cause but maybe not always an achievable goal.

As such, this change muddles what the actual recommendations are.

File: scaling_performance/optimizing_storage.adoc [hunk 5]
```diff
@@ -161,9 +179,16 @@ In an {product-title} hosted logging cluster deployment:
 * The preferred storage technology is block storage.
 * It is not recommended to use NAS storage (excluding {gluster-native}/{gluster-external} GlusterFS as it uses a block storage interface from iSCSI) for a hosted metrics cluster deployment with production workloads.
 
-[WARNING]
+[IMPORTANT]
 ====
-Corruption may occur when using NFS to back hosted logging with production workloads.
+Testing has shown issues with using the NFS server on RHEL as storage backend for
+the container registry. This includes the embedded registry and Quay, Cassandra
```
Comment (lpsantil) [inline scaling_performance/optimizing_storage.adoc]: Do you mean OpenShift Container Registry when you say `embedded registry`?

File: install/configuring_inventory_file.adoc [hunk 6]
```diff
@@ -1609,6 +1611,23 @@ persistent storage for logging. If `openshift_logging_es_pvc_dynamic` is
 not set, then cluster logging data is stored in an `emptyDir` volume, which will
 be deleted when the Elasticsearch pod terminates.
 
+[IMPORTANT]
+====
+Testing has shown issues with using the NFS server on RHEL as storage backend
+for the container registry. This includes the embedded registry and Quay,
+Cassandra for metrics storage, and ElasticSearch for logging storage. Therefore,
+using NFS on RHEL as storage for these framework services is not recommended.
```
Comment (lpsantil) [inline install/configuring_inventory_file.adoc]: It's not just `NFS on RHEL` that can have issues.

**Review by lpsantil — CHANGES_REQUESTED** [review]
Overall, I don't like the blanket copy & paste statements.  `NFS on RHEL` is not the only NFS server with issues.  But it is not a server X vs. server Y issue; it is a protocol and application architecture issue.

The `Optimizing Persistent Storage` doc page tries to make it clear.  Block for DB and apps that need it.  File for Apps that don't need Read-After-Write Consistency.  Block for Metrics, Logging.  Object for Registry or Block if you're not going to scale out the registry.

@cpitman @raffaelespazzoli @sabre1041  What do you think?

**Comment by bmcelvee** [comment]
Preview build: http://file.rdu.redhat.com/~bmcelvee/18092018%5eC/install/configuring_inventory_file.html

**Comment by emmahone** [comment]
I just reviewed these docs and here are my thoughts on changes to be made. 

~~~
https://github.com/openshift/openshift-docs/pull/11947/commits/0b7afd221a950b0d58cc6653b790c9a442bac864#diff-80f7c8ee1abcb1ceb42d8350b49bf604R1483

	-Change framework to core

https://github.com/openshift/openshift-docs/pull/11947/commits/0b7afd221a950b0d58cc6653b790c9a442bac864#diff-80f7c8ee1abcb1ceb42d8350b49bf604R1492

	-Change framework to core

https://github.com/openshift/openshift-docs/pull/11947/commits/0b7afd221a950b0d58cc6653b790c9a442bac864#diff-80f7c8ee1abcb1ceb42d8350b49bf604R1485

	-The reason we don't recommend NFS or NAS/SAN storage is because of IO constraints (cassandra write and reads a ton of data) rather than having the readwriteonce access mode. It is a problem, but not the biggest reason we don't want file based network storage. 

These statements throughout the docs could be modified: "Therefore, using NFS on RHEL as storage for these framework services is not recommended."

	We do not recommend using NFS to back PVs used by core services. 

As someone who had 0 idea what anti-pattern means, can we dumb that down to my level =]
~~~

**Comment by bmcelvee** [comment]
@openshift/team-documentation PTAL

File: install/configuring_inventory_file.adoc [hunk 7]
```diff
@@ -1002,18 +1002,25 @@ can be deployed using the installer.
 
 If no registry storage options are used, the default OpenShift Container
 Registry is ephemeral and all data will be lost when the pod no longer exists.
+
+[IMPORTANT]
+====
+Testing has shown issues with using the NFS server on RHEL as storage backend
```
Comment (ahardin-rh) [inline install/configuring_inventory_file.adoc]: s/Testing has shown/Testing shows

File: install/configuring_inventory_file.adoc [hunk 8]
```diff
@@ -1002,18 +1002,25 @@ can be deployed using the installer.
...
+for the container registry. This includes the OpenShift Container Registry and Quay.
+Therefore, using NFS to back PVs used by core services is not recommended.
+
+Other NFS implementations on the marketplace may not have these issues. Please
```
Comment (ahardin-rh) [inline install/configuring_inventory_file.adoc]: s/Please contact/Contact

Comment (kalexand-rh) [reply install/configuring_inventory_file.adoc]: s/may/might

File: install/configuring_inventory_file.adoc [hunk 9]
```diff
@@ -1002,18 +1002,25 @@ can be deployed using the installer.
...
+contact the individual NFS implementation vendor for more information on any
+testing they may have performed against these OpenShift core components.
```
Comment (ahardin-rh) [inline install/configuring_inventory_file.adoc]: s/testing they may have performed/testing that was possibly completed

**Comment by jhoyt-rh** [comment]
+
+[IMPORTANT]
+====
+Testing has shown issues with using the NFS server on RHEL as storage backend
+for the container registry. This includes the OpenShift Container
Registry and Quay.
+Therefore, using NFS to back PVs used by core services is not recommended.

s/Therefore, we do not recommend using NFS to back PVs used by core
services.

(per meeting discussion with Evan yesterday)

Joan Hoyt
Sr. Technical Writer, Customer Content Services/OpenShift
Red Hat, Inc
Westford

**Comment by jhoyt-rh** [comment]
+Testing has shown issues with using the NFS server on RHEL as storage backend
+for the container registry. This includes the OpenShift Container
Registry and Quay.
+Therefore, using NFS to back PVs used by core services is not recommended.

s/Therefore, we recommend that you do not use NFS to back PVs used by core
services.

(Sorry, made error above when referring to meeting discussion with Evan)


Joan Hoyt
Sr. Technical Writer, Customer Content Services/OpenShift
Red Hat, Inc
Westford
